File: app_canibuythat/src/main/java/com/gb/canibuythat/fcm/MonzoDispatchInstanceIdService.java
package com.gb.canibuythat.fcm;

import android.util.Log;

import com.google.firebase.iid.FirebaseInstanceId;
import com.google.firebase.iid.FirebaseInstanceIdService;

public class MonzoDispatchInstanceIdService extends FirebaseInstanceIdService {

    private static final String TAG = "MonzoDispatch";

    @Override
    public void onTokenRefresh() {
        String refreshedToken = FirebaseInstanceId.getInstance().getToken();
        Log.d(TAG, "Refreshed token: " + refreshedToken);

        sendRegistrationToServer(refreshedToken);
    }

    /**
     * Persist token to third-party servers.
     *
     * @param token The new token.
     */
    private void sendRegistrationToServer(String token) {
        // TODO: Implement this method to send token to your app server.
    }
}

File: app_canibuythat/src/main/java/com/gb/canibuythat/fcm/MonzoDispatchInstanceIdService.kt
package com.gb.canibuythat.fcm

import android.content.Context
import android.util.Log
import android.widget.Toast
import com.gb.canibuythat.CredentialsProvider
import com.gb.canibuythat.di.Injector
import com.gb.canibuythat.exception.ErrorHandler
import com.gb.canibuythat.presenter.MonzoDispatchPresenter
import com.gb.canibuythat.screen.MonzoDispatchScreen
import com.google.firebase.iid.FirebaseInstanceId
import com.google.firebase.iid.FirebaseInstanceIdService
import javax.inject.Inject

class MonzoDispatchInstanceIdService : FirebaseInstanceIdService(), MonzoDispatchScreen {

    @field:[Inject] lateinit var errorHandler: ErrorHandler
    @field:[Inject] lateinit var appContext: Context
    @field:[Inject] lateinit var presenter: MonzoDispatchPresenter
    @field:[Inject] lateinit var credentialsProvider: CredentialsProvider

    init {
        Injector.INSTANCE.graph.inject(this)
        presenter.setScreen(this)
    }

    override fun onTokenRefresh() {
        if (!credentialsProvider.accessToken.isNullOrEmpty() || credentialsProvider.isRefresh()) {
            val refreshedToken = FirebaseInstanceId.getInstance().token
            Log.d(TAG, "Refreshed token: " + refreshedToken!!)
            presenter.sendFCMTokenToServer(refreshedToken)
        }
    }

    override fun showProgress() {
    }

    override fun showToast(message: String) {
        Toast.makeText(appContext, message, Toast.LENGTH_SHORT).show()
    }

    override fun hideProgress() {
    }

    companion object {
        private val TAG = "MonzoDispatch"
    }
}
